File: src/test/java/no/nav/fo/veilarbdialog/rest/DialogBeskjedTest.java
package no.nav.fo.veilarbdialog.rest;

import io.restassured.RestAssured;
import lombok.extern.slf4j.Slf4j;
import net.javacrumbs.shedlock.core.LockProvider;
import net.javacrumbs.shedlock.provider.jdbctemplate.JdbcTemplateLockProvider;
import no.nav.brukernotifikasjon.schemas.input.BeskjedInput;
import no.nav.brukernotifikasjon.schemas.input.DoneInput;
import no.nav.brukernotifikasjon.schemas.input.NokkelInput;
import no.nav.fo.veilarbdialog.brukernotifikasjon.*;
import no.nav.fo.veilarbdialog.brukernotifikasjon.entity.BrukernotifikasjonEntity;
import no.nav.fo.veilarbdialog.domain.DialogDTO;
import no.nav.fo.veilarbdialog.domain.NyHenvendelseDTO;
import no.nav.fo.veilarbdialog.mock_nav_modell.BrukerOptions;
import no.nav.fo.veilarbdialog.mock_nav_modell.MockBruker;
import no.nav.fo.veilarbdialog.mock_nav_modell.MockNavService;
import no.nav.fo.veilarbdialog.mock_nav_modell.MockVeileder;
import no.nav.fo.veilarbdialog.service.ScheduleRessurs;
import no.nav.fo.veilarbdialog.util.KafkaTestService;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.assertj.core.api.SoftAssertions;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.cloud.contract.wiremock.AutoConfigureWireMock;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.kafka.test.utils.KafkaTestUtils;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.context.junit4.SpringRunner;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Date;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.context.jdbc.Sql.ExecutionPhase.BEFORE_TEST_METHOD;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@RunWith(SpringRunner.class)
@AutoConfigureWireMock(port = 0)
@Slf4j
@Sql(
        scripts = "/db/testdata/slett_alle_dialoger.sql",
        executionPhase = BEFORE_TEST_METHOD
)
public class DialogBeskjedTest {

    @LocalServerPort
    protected int port;

    @Autowired
    ScheduleRessurs scheduleRessurs;

    @Autowired
    BrukernotifikasjonRepository brukernotifikasjonRepository;

    @Autowired
    JdbcTemplate jdbcTemplate;

    @Autowired
    LockProvider lockProvider;

    @Autowired
    KafkaTestService kafkaTestService;

    @Autowired
    BrukernotifikasjonService brukernotifikasjonService;

    @Value("${application.topic.ut.brukernotifikasjon.beskjed}")
    private String brukernotifikasjonBeskjedTopic;

    @Value("${application.topic.ut.brukernotifikasjon.done}")
    private String brukernotifikasjonDoneTopic;

    Consumer<NokkelInput, BeskjedInput> brukerNotifikasjonBeskjedConsumer;

    Consumer<NokkelInput, DoneInput> brukerNotifikasjonDoneConsumer;

    @Before
    public void setup() {
        JdbcTemplateLockProvider l = (JdbcTemplateLockProvider) lockProvider;
        l.clearCache();
        RestAssured.port = port;
        brukerNotifikasjonBeskjedConsumer = kafkaTestService.createAvroAvroConsumer(brukernotifikasjonBeskjedTopic);
        brukerNotifikasjonDoneConsumer = kafkaTestService.createAvroAvroConsumer(brukernotifikasjonDoneTopic);
    }

    @Test
    public void beskjed_happy_case() {
        MockBruker mockBruker = MockNavService.createHappyBruker();
        MockVeileder mockVeileder = MockNavService.createVeileder(mockBruker);

        DialogDTO dialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst").setOverskrift("overskrift"))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);

        // Setter sendt til å være 1 sekund tidligere pga. grace period
        settHenvendelseSendtForNSekundSiden(dialog.getHenvendelser().get(0).getId(), 1);

        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();

        ConsumerRecord<NokkelInput, BeskjedInput> brukernotifikasjonRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonBeskjedConsumer, brukernotifikasjonBeskjedTopic, 5000L);

        assertThat(brukernotifikasjonRecord.value().getTekst()).isEqualTo(BrukernotifikasjonTekst.BESKJED_BRUKERNOTIFIKASJON_TEKST);

        BrukernotifikasjonEntity brukernotifikasjonEntity =
                brukernotifikasjonRepository.hentBrukernotifikasjonForDialogId(Long.parseLong(dialog.getId()), BrukernotifikasjonsType.BESKJED).get(0);

        SoftAssertions.assertSoftly(
                assertions -> {
                    assertions.assertThat(brukernotifikasjonEntity.dialogId()).isEqualTo(Long.valueOf(dialog.getId()));
                    assertions.assertThat(brukernotifikasjonEntity.type()).isEqualTo(BrukernotifikasjonsType.BESKJED);
                    assertions.assertThat(brukernotifikasjonEntity.status()).isEqualTo(BrukernotifikasjonBehandlingStatus.SENDT);
                }
        );

        mockBruker.createRequest()
                .put("/veilarbdialog/api/dialog/{dialogId}/les", dialog.getId())
                .then()
                .statusCode(200);

        brukernotifikasjonService.sendDoneBrukernotifikasjoner();

        ConsumerRecord<NokkelInput, DoneInput> doneRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonDoneConsumer, brukernotifikasjonDoneTopic, 5000L);

        NokkelInput nokkel = doneRecord.key();

        assertThat(mockBruker.getFnr()).isEqualTo(nokkel.getFodselsnummer());
    }

    @Test
    public void ikke_beskjed_foer_grace_periode_utlopt() {
        MockBruker mockBruker = MockNavService.createHappyBruker();
        MockVeileder mockVeileder = MockNavService.createVeileder(mockBruker);

        mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst").setOverskrift("overskrift"))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);

        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();

        boolean harKonsumertAlleMeldinger = kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonBeskjedTopic, brukerNotifikasjonBeskjedConsumer);
        assertThat(harKonsumertAlleMeldinger).isTrue();
    }

    @Test
    public void ikke_beskjed_naar_maxalder_er_passert() {
        MockBruker mockBruker = MockNavService.createHappyBruker();
        MockVeileder mockVeileder = MockNavService.createVeileder(mockBruker);

        DialogDTO dialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst").setOverskrift("overskrift"))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);

        // Setter sendt til å være 3 sekund tidligere pga. max alder
        settHenvendelseSendtForNSekundSiden(dialog.getHenvendelser().get(0).getId(), 3);

        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();

        boolean harKonsumertAlleMeldinger = kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonBeskjedTopic, brukerNotifikasjonBeskjedConsumer);
        assertThat(harKonsumertAlleMeldinger).isTrue();
    }

    @Test
    public void kan_ikke_varsles() {
        MockBruker bruker = MockNavService.createHappyBruker();
        BrukerOptions reservertKrr = bruker.getBrukerOptions().toBuilder().erReservertKrr(true).build();
        MockNavService.updateBruker(bruker, reservertKrr);

        MockVeileder veileder = MockNavService.createVeileder(bruker);

        veileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst").setOverskrift("overskrift"))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", bruker.getAktorId())
                .then()
                .statusCode(409);

    }

    @Test
    public void ikkeDobleNotifikasjonerPaaSammeDialog() {

        MockBruker mockBruker = MockNavService.createHappyBruker();
        MockVeileder mockVeileder = MockNavService.createVeileder(mockBruker);

        DialogDTO dialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst").setOverskrift("overskrift"))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);

        // Hy henvendelse samme dialog
        DialogDTO sammeDialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst2").setDialogId(dialog.getId()))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);

        // Setter sendt til å være 1 sekund tidligere pga. grace period
        settHenvendelseSendtForNSekundSiden(sammeDialog.getHenvendelser().get(0).getId(), 1);
        settHenvendelseSendtForNSekundSiden(sammeDialog.getHenvendelser().get(1).getId(), 1);

        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();

        ConsumerRecord<NokkelInput, BeskjedInput> brukernotifikasjonRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonBeskjedConsumer, brukernotifikasjonBeskjedTopic, 20000L);

        Assert.assertTrue(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonBeskjedTopic, brukerNotifikasjonBeskjedConsumer));

        assertThat(brukernotifikasjonRecord.value().getTekst()).isEqualTo(BrukernotifikasjonTekst.BESKJED_BRUKERNOTIFIKASJON_TEKST);




        BrukernotifikasjonEntity brukernotifikasjonEntity =
                brukernotifikasjonRepository.hentBrukernotifikasjonForDialogId(Long.parseLong(dialog.getId()), BrukernotifikasjonsType.BESKJED).get(0);

        SoftAssertions.assertSoftly(
                assertions -> {
                    assertions.assertThat(brukernotifikasjonEntity.dialogId()).isEqualTo(Long.valueOf(dialog.getId()));
                    assertions.assertThat(brukernotifikasjonEntity.type()).isEqualTo(BrukernotifikasjonsType.BESKJED);
                    assertions.assertThat(brukernotifikasjonEntity.status()).isEqualTo(BrukernotifikasjonBehandlingStatus.SENDT);
                }
        );

        mockBruker.createRequest()
                .put("/veilarbdialog/api/dialog/{dialogId}/les", dialog.getId())
                .then()
                .statusCode(200);

        brukernotifikasjonService.sendDoneBrukernotifikasjoner();

        ConsumerRecord<NokkelInput, DoneInput> doneRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonDoneConsumer, brukernotifikasjonDoneTopic, 10000L);

        Assert.assertTrue(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonDoneTopic, brukerNotifikasjonDoneConsumer));

        NokkelInput nokkel = doneRecord.key();

        assertThat(mockBruker.getFnr()).isEqualTo(nokkel.getFodselsnummer());

    }

    @Test
    public void ingenDobleNotifikasjonerPaaSammeDialogMedIntervall() {

        MockBruker mockBruker = MockNavService.createHappyBruker();
        MockVeileder mockVeileder = MockNavService.createVeileder(mockBruker);

        DialogDTO dialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst").setOverskrift("overskrift"))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);



        // Setter sendt til å være 1 sekund tidligere pga. grace period
        settHenvendelseSendtForNSekundSiden(dialog.getHenvendelser().get(0).getId(), 1);

        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();

        ConsumerRecord<NokkelInput, BeskjedInput> brukernotifikasjonRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonBeskjedConsumer, brukernotifikasjonBeskjedTopic, 20000L);

        Assert.assertTrue(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonBeskjedTopic, brukerNotifikasjonBeskjedConsumer));

        assertThat(brukernotifikasjonRecord.value().getTekst()).isEqualTo(BrukernotifikasjonTekst.BESKJED_BRUKERNOTIFIKASJON_TEKST);


        // Hy henvendelse samme dialog
        DialogDTO sammedialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst2").setDialogId(dialog.getId()))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);


        // Setter sendt til å være 1 sekund tidligere pga. grace period
        settHenvendelseSendtForNSekundSiden(sammedialog.getHenvendelser().get(1).getId(), 1);

        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();

        // Ingen nye beskjeder siden forrige henvendelse ikke er lest

        Assert.assertTrue(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonBeskjedTopic, brukerNotifikasjonBeskjedConsumer));

        mockBruker.createRequest()
                .put("/veilarbdialog/api/dialog/{dialogId}/les", dialog.getId())
                .then()
                .statusCode(200);

        brukernotifikasjonService.sendDoneBrukernotifikasjoner();

        ConsumerRecord<NokkelInput, DoneInput> doneRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonDoneConsumer, brukernotifikasjonDoneTopic, 10000L);

        Assert.assertTrue(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonDoneTopic, brukerNotifikasjonDoneConsumer));

        NokkelInput nokkel = doneRecord.key();

        assertThat(mockBruker.getFnr()).isEqualTo(nokkel.getFodselsnummer());

    }


    @Test
    public void nyNotifikasjonPaaSammeDialogNaarFoersteErLest() throws InterruptedException {

        MockBruker mockBruker = MockNavService.createHappyBruker();
        MockVeileder mockVeileder = MockNavService.createVeileder(mockBruker);

        DialogDTO dialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst").setOverskrift("overskrift"))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);

        settHenvendelseSendtForNSekundSiden(dialog.getHenvendelser().get(0).getId(), 1);
        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();


        ConsumerRecord<NokkelInput, BeskjedInput> brukernotifikasjonRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonBeskjedConsumer, brukernotifikasjonBeskjedTopic, 10000L);

        Assert.assertTrue(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonBeskjedTopic, brukerNotifikasjonBeskjedConsumer));

        assertThat(brukernotifikasjonRecord.value().getTekst()).isEqualTo(BrukernotifikasjonTekst.BESKJED_BRUKERNOTIFIKASJON_TEKST);

        mockBruker.createRequest()
                .put("/veilarbdialog/api/dialog/{dialogId}/les", dialog.getId())
                .then()
                .statusCode(200);

        brukernotifikasjonService.sendDoneBrukernotifikasjoner();

        ConsumerRecord<NokkelInput, DoneInput> doneRecord =
                KafkaTestUtils.getSingleRecord(brukerNotifikasjonDoneConsumer, brukernotifikasjonDoneTopic, 10000L);

        Assert.assertTrue(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonDoneTopic, brukerNotifikasjonDoneConsumer));



        // Hy henvendelse samme dialog
        DialogDTO sammeDialog = mockVeileder.createRequest()
                .body(new NyHenvendelseDTO().setTekst("tekst2").setDialogId(dialog.getId()))
                .post("/veilarbdialog/api/dialog?aktorId={aktorId}", mockBruker.getAktorId())
                .then()
                .statusCode(200)
                .extract()
                .as(DialogDTO.class);

        Thread.sleep(1000L);
        scheduleRessurs.sendBrukernotifikasjonerForUlesteDialoger();
        brukernotifikasjonService.sendPendingBrukernotifikasjoner();

        // Ny beskjed siden forrige henvendelse er lest

        Assert.assertFalse(kafkaTestService.harKonsumertAlleMeldinger(brukernotifikasjonBeskjedTopic, brukerNotifikasjonBeskjedConsumer));


    }

    // Setter sendt til å være N sekund tidligere pga. grace period
    private void settHenvendelseSendtForNSekundSiden(String henvendelseId, int sekunderSiden) {
        Date nySendt = Date.from(Instant.now().minus(sekunderSiden, ChronoUnit.SECONDS));
        int update = jdbcTemplate.update("""
                                UPDATE HENVENDELSE
                                SET SENDT = ?
                                WHERE HENVENDELSE_ID = ?
                        """,
                nySendt,
                henvendelseId
        );
        assertThat(update).isEqualTo(1);
    }

}

File: src/test/java/no/nav/fo/veilarbdialog/rest/DialogBeskjedTest.kt
package no.nav.fo.veilarbdialog.rest

import no.nav.common.json.JsonUtils
import no.nav.fo.veilarbdialog.SpringBootTestBase
import no.nav.fo.veilarbdialog.domain.DialogDTO
import no.nav.fo.veilarbdialog.mock_nav_modell.BrukerOptions
import no.nav.fo.veilarbdialog.mock_nav_modell.MockNavService
import no.nav.fo.veilarbdialog.util.KafkaTestService
import no.nav.tms.varsel.action.OpprettVarsel
import org.apache.kafka.clients.consumer.Consumer
import org.assertj.core.api.Assertions
import org.assertj.core.api.SoftAssertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.kafka.test.utils.KafkaTestUtils
import java.sql.Types
import java.time.Instant
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import no.nav.fo.veilarbdialog.brukernotifikasjon.BrukernotifikasjonBehandlingStatus.SENDT
import no.nav.fo.veilarbdialog.brukernotifikasjon.BrukernotifikasjonTekst.NY_MELDING_TEKST
import no.nav.fo.veilarbdialog.brukernotifikasjon.BrukernotifikasjonsType.BESKJED
import no.nav.fo.veilarbdialog.minsidevarsler.MinsideVarselService
import no.nav.fo.veilarbdialog.domain.NyMeldingDTO
import no.nav.fo.veilarbdialog.minsidevarsler.dto.MinsideVarselDao
import no.nav.fo.veilarbdialog.mock_nav_modell.MockBruker
import no.nav.fo.veilarbdialog.mock_nav_modell.MockVeileder
import no.nav.tms.varsel.action.InaktiverVarsel
import no.nav.tms.varsel.action.Varseltype
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Assertions.assertTrue
import java.time.temporal.ChronoUnit
import java.util.Date

internal class DialogBeskjedTest(
    @Autowired
    val minsideVarselDao: MinsideVarselDao,
    @Autowired
    var kafkaTestService: KafkaTestService,
    @Autowired
    var minsideVarselService: MinsideVarselService,
    @Value("\${application.topic.ut.minside.varsel}")
    private val minsideVarselTopic: String,
) : SpringBootTestBase() {

    init {
        JsonUtils.getMapper().registerKotlinModule()
    }

    var minsideVarselConsumer: Consumer<String, String>? = null

    @BeforeEach
    fun setupL() {
        minsideVarselConsumer = kafkaTestService.createStringStringConsumer(minsideVarselTopic)
    }

    private fun assertOpprettetVarselPublisertPåKafka(): OpprettVarsel {
        val brukernotifikasjonRecord =
            KafkaTestUtils.getSingleRecord<String, String>(
                minsideVarselConsumer,
                minsideVarselTopic,
                KafkaTestService.DEFAULT_WAIT_TIMEOUT
            )
        assertTrue(
            kafkaTestService.harKonsumertAlleMeldinger(
                minsideVarselTopic,
                minsideVarselConsumer,
            ), "Forventet at det ikke skulle være flere meldinger på minsideVarselTopic"
        )
        val opprettVarsel =
            JsonUtils.fromJson<OpprettVarsel>(brukernotifikasjonRecord.value(), OpprettVarsel::class.java)
        return opprettVarsel
    }

    private fun assertInaktiveringPublisertPåKafka(): InaktiverVarsel {
        val doneRecord = KafkaTestUtils.getSingleRecord<String, String>(
                minsideVarselConsumer,
                minsideVarselTopic,
                KafkaTestService.DEFAULT_WAIT_TIMEOUT
            )
        assertTrue(
            kafkaTestService.harKonsumertAlleMeldinger(
                minsideVarselTopic,
                minsideVarselConsumer,
            )
        )
        return JsonUtils.fromJson(doneRecord.value(), InaktiverVarsel::class.java)
    }

    @Test
    fun `varsel skal opprettes når veileder sender melding og inaktiveres når bruker leser melding`() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        val dialog = mockVeileder.sendEnMelding(mockBruker)

        minsideVarselService.sendPendingVarslerCronImpl()
        val opprettVarsel = assertOpprettetVarselPublisertPåKafka()

        assertThat(opprettVarsel.tekster.first().tekst).isEqualTo(NY_MELDING_TEKST)
        assertThat(opprettVarsel.type).isEqualTo(Varseltype.Beskjed)

        val brukernotifikasjonEntity = minsideVarselDao.getVarslerForDialog(
            dialog.id.toLong(),
        )[0]

        assertSoftly { assertions ->
//            assertions.assertThat(brukernotifikasjonEntity.dialogId).isEqualTo(dialog.id.toLong())
            assertions.assertThat(brukernotifikasjonEntity.status).isEqualTo(SENDT)
        }

        mockBruker.lesMelding(dialog.id)
        minsideVarselService.sendInktiveringPåKafkaPåVarslerSomSkalAvsluttes()

        assertInaktiveringPublisertPåKafka()
        Assertions.assertThat(mockBruker.fnr).isEqualTo(opprettVarsel.ident)
    }

    private fun MockVeileder.sendEnMelding(mockBruker: MockBruker, dialogId: String? = null): DialogDTO {
        return this.createRequest()
            .body(NyMeldingDTO().setTekst("tekst").setOverskrift("overskrift").setDialogId(dialogId))
            .post("/veilarbdialog/api/dialog?fnr={fnr}", mockBruker.fnr)
            .then()
            .statusCode(200)
            .extract()
            .`as`<DialogDTO>(DialogDTO::class.java)
    }

    private fun MockBruker.lesMelding(dialogId: String) {
        this.createRequest()
            .put("/veilarbdialog/api/dialog/{dialogId}/les", dialogId)
            .then()
            .statusCode(200)
    }

    @Test
    fun sending_av_melding_skal_putte_varsel_i_pending() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        mockVeileder.sendEnMelding(mockBruker)
        val sendteMeldinger = minsideVarselService.sendPendingVarslerCronImpl()
        assertThat(sendteMeldinger).isEqualTo(1)
        val varselOpprettelse = assertOpprettetVarselPublisertPåKafka()
        assertThat(varselOpprettelse.eksternVarsling?.kanBatches).isTrue()
    }

    @Test
    fun kan_ikke_varsles() {
        val bruker = MockNavService.createBruker(BrukerOptions.happyBruker().toBuilder().erReservertKrr(true).build())
        val veileder = MockNavService.createVeileder(bruker)

        veileder.createRequest()
            .body(NyMeldingDTO().setTekst("tekst").setOverskrift("overskrift"))
            .post("/veilarbdialog/api/dialog?fnr={fnr}", bruker.fnr)
            .then()
            .statusCode(409)
    }

    @Test
    fun ikkeDobleNotifikasjonerPaaSammeDialog() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        val dialog = mockVeileder.sendEnMelding(mockBruker)
        // Hy henvendelse samme dialog
        mockVeileder.sendEnMelding(mockBruker, dialog.id)

        val sendteVarsler = minsideVarselService.sendPendingVarslerCronImpl()
        assertThat(sendteVarsler).isEqualTo(1).withFailMessage("Skal sende ut 1 varsel som stod i PENDING")
        val opprettVarsel = assertOpprettetVarselPublisertPåKafka()
        assertThat(opprettVarsel.tekster.first().tekst).isEqualTo(NY_MELDING_TEKST)

        val brukernotifikasjonEntity = minsideVarselDao.getVarslerForDialog(dialog.id.toLong(),)[0]

        assertSoftly {
            assertions: SoftAssertions ->
//            assertions.assertThat(brukernotifikasjonEntity.dialogId).isEqualTo(dialog.id.toLong())
//            assertions.assertThat(brukernotifikasjonEntity.type).isEqualTo(BESKJED)
            assertions.assertThat(brukernotifikasjonEntity.status).isEqualTo(SENDT)
        }

        mockBruker.lesMelding(dialog.id)

        minsideVarselService.sendInktiveringPåKafkaPåVarslerSomSkalAvsluttes()
        val inaktivering = assertInaktiveringPublisertPåKafka()
        assertThat(opprettVarsel.varselId).isEqualTo(inaktivering.varselId)
    }

    @Test
    fun `ingen dobble minsiede-varsler på samme dialog med intervall`() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        // Send 2 meldinger på samme dialog
        val dialog = mockVeileder.sendEnMelding(mockBruker)
        mockVeileder.sendEnMelding(mockBruker, dialogId = dialog.id)

        val sendteVarsler = minsideVarselService.sendPendingVarslerCronImpl()
        assertThat(sendteVarsler).isEqualTo(1).withFailMessage("Skal sende 1 og bare 1 varsel")
        assertOpprettetVarselPublisertPåKafka()
    }

    @Test
    fun `les skal inaktivere varsel selvom varsel ikke ble publisert på kafka`() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        val dialog = mockVeileder.sendEnMelding(mockBruker) // PENDING
        mockBruker.lesMelding(dialog.id) // PENDING -> SKAL_AVSLUTTES
        minsideVarselService.sendInktiveringPåKafkaPåVarslerSomSkalAvsluttes() // SKAL_AVSLUTTES -> AVSLUTTET

        assertInaktiveringPublisertPåKafka()
    }

    @Test
    @Throws(InterruptedException::class)
    fun `flere meldinger på samme dialog skal ikke sende flere notifikasjoner hvis det er under 30 min siden forrige varsel`() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        val dialog = mockVeileder.sendEnMelding(mockBruker)
        minsideVarselService.sendPendingVarslerCronImpl()

        val opprettVarsel = assertOpprettetVarselPublisertPåKafka()
        assertThat(opprettVarsel.tekster.first().tekst).isEqualTo(NY_MELDING_TEKST)

        // Hy henvendelse samme dialog
        mockVeileder.sendEnMelding(mockBruker, dialog.id)
        minsideVarselService.sendPendingVarslerCronImpl()

        assertTrue(
            kafkaTestService.harKonsumertAlleMeldinger(
                minsideVarselTopic,
                minsideVarselConsumer,
            )
        )
    }

    @Test
    @Throws(InterruptedException::class)
    fun `flere meldinger på samme dialog skal sende flere notifikasjoner hvis det er overt 30 min siden forrige varsel`() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        val dialog = mockVeileder.sendEnMelding(mockBruker)
        settBrukernotifikasjonOpprettetForNMinuttSiden(dialog.id, 31)
        minsideVarselService.sendPendingVarslerCronImpl()

        val opprettVarsel = assertOpprettetVarselPublisertPåKafka()
        assertThat(opprettVarsel.tekster.first().tekst).isEqualTo(NY_MELDING_TEKST)

        // Hy henvendelse samme dialog
        mockVeileder.sendEnMelding(mockBruker, dialog.id)
        minsideVarselService.sendPendingVarslerCronImpl()

        assertOpprettetVarselPublisertPåKafka()
    }

    @Test
    @Throws(InterruptedException::class)
    fun `flere meldinger på samme dialog når første er lest skal sende notifikasjon selvom det er under 30 min siden`() {
        val mockBruker = MockNavService.createHappyBruker()
        val mockVeileder = MockNavService.createVeileder(mockBruker)

        val dialog = mockVeileder.sendEnMelding(mockBruker)
        settBrukernotifikasjonOpprettetForNMinuttSiden(dialog.id, 1)
        minsideVarselService.sendPendingVarslerCronImpl()

        val opprettVarsel = assertOpprettetVarselPublisertPåKafka()
        assertThat(opprettVarsel.tekster.first().tekst).isEqualTo(NY_MELDING_TEKST)

        mockBruker.lesMelding(dialog.id)
        minsideVarselService.sendInktiveringPåKafkaPåVarslerSomSkalAvsluttes()

        assertInaktiveringPublisertPåKafka()

        // Hy henvendelse samme dialog
        mockVeileder.sendEnMelding(mockBruker, dialog.id)
        minsideVarselService.sendPendingVarslerCronImpl()

        assertOpprettetVarselPublisertPåKafka()
    }

    private fun settBrukernotifikasjonOpprettetForNMinuttSiden(dialogId: String?, minutterSiden: Int) {
        val namedJdbcTemplate = NamedParameterJdbcTemplate(jdbcTemplate)
        val justertOpprettet = Date.from(Instant.now().minus(minutterSiden.toLong(), ChronoUnit.MINUTES))

        val update = namedJdbcTemplate.update(
            """
                UPDATE min_side_varsel 
                SET opprettet = :sendt
                FROM min_side_varsel_dialog_mapping 
                WHERE min_side_varsel_dialog_mapping.dialog_id = :dialogId
                    AND min_side_varsel_dialog_mapping.varsel_id = min_side_varsel.varsel_id
                        
                        """.trimIndent(),
            MapSqlParameterSource("sendt", justertOpprettet)
                .addValue("dialogId", dialogId, Types.BIGINT)
        )
        assertThat(update).isEqualTo(1)
    }
}
